src/gtimelog/main.py

```diff
@@ -172,6 +174,143 @@ def make_option(long_name, short_name=None, flags=0, arg=GLib.OptionArg.NONE,
     return option
 
 
+# Global HTTP stuff
+
+class Authenticator(object):
+    # try to use GNOME Keyring if available
+    try:
+        gi.require_version('GnomeKeyring', '1.0')
+        from gi.repository import GnomeKeyring as gnomekeyring
+    except (ValueError, ImportError):
+        gnomekeyring = None
+
+    def __init__(self):
+        object.__init__(self)
+        self.pending = []
+        self.lookup_in_progress = False
+        self.username = None
+        self.password = None
+
+    def find_in_keyring(self, uri, callback):
+        """Attempts to load a username and password from the keyring, if the
+        keyring is available"""
+        username = self.username
+        password = self.password
+
+        if self.gnomekeyring is None:
+            callback(username, password)
+            return
+
+        # FIXME - would be nice to make all keyring calls async, to dodge
+        # the possibility of blocking the UI. The code is all set up for
+        # that, but there's no easy way to use the keyring asynchronously
+        # from Python (as of Gnome 3.2)...
+        l = self.gnomekeyring.find_network_password_sync (
+                None,           # user
+                uri.get_host(), # domain
+                uri.get_host(), # server
+                None,           # object
+                uri.get_scheme(),# protocol
+                None,           # authtype
+                uri.get_port()) # port
+
+        result, keys = l
+        if result == self.gnomekeyring.Result.NO_MATCH:
+            # We didn't find any passwords, just continue
+            pass
+        elif result == self.gnomekeyring.Result.NO_KEYRING_DAEMON:
+            pass
+        else:
+            entry = keys[-1] # take the last key (Why?)
+            username = entry.user
+            password = entry.password
+
+        callback(username, password)
+
+    def save_to_keyring(self, uri, username, password):
+        self.gnomekeyring.set_network_password_sync (
+                None,           # keyring
+                username,       # user
+                uri.get_host(), # domain
+                uri.get_host(), # server
+                None,           # object
+                uri.get_scheme(),# protocol
+                None,           # authtype
+                uri.get_port(), # port
+                password)       # password
+
+    def ask_the_user(self, auth, uri, callback):
+        """Pops up a username/password dialog for uri"""
+        mountoperation = Gtk.MountOperation.new()
+
+        def on_reply(m, r):
+            save_to_keyring = self.gnomekeyring and (m.get_password_save() == Gio.PasswordSave.PERMANENTLY)
+
+            if r == Gio.MountOperationResult.HANDLED:
+                username = m.get_username()
+                password = m.get_password()
+
+                if username and password:
+                    if self.gnomekeyring and (m.get_password_save() == Gio.PasswordSave.PERMANENTLY):
```

Comment:
You already have this subexpression stored in `save_to_keyring`.  Reuse?

Reply:
Better drop :)